File: mpi/practicals/helloworld/helloworld.c
#include <stdio.h>
#include "/* INSERT MISSING HEADER */"

///////////////////
// MPI Hello World
///////////////////
//
// Classical example printing "Hello World" in several processes
//
// example usage:
//		compile: mpicc -o helloworld helloworld.c
//		run: mpirun -n 4 helloworld
//

int main (int argc, char *argv[]) {

    double my_data;

/* Turn it into an MPI program (initialise MPI) */   
   ; /* <-- INSERT MISSING MPI FUNCTION ON THIS LINE */
   ; /* <-- INSERT Get the rank within the global communicator */
   ; /* <-- INSERT Get the total number of ranks within the global communicator */

/* Experiment with Hello message in all processes and only in the master process */

   printf("Hello world from rank %d of size %d \n"/*, INSERT VARIABLES */);
   if (/*INSERT CONDITION*/) { /* Only rank == 0 should print */
     printf("Hello world from the master process rank %d of size %d \n"/*, INSERT VARIABLES */);
   }
   fflush(stdout);  // Flush output to ensure it appears immediately
   
/* Properly shutdown MPI (finalise MPI) */
   ; /* <-- INSERT MISSING MPI FUNCTION ON THIS LINE */
}

File: mpi/practicals/helloworld/helloworld.f90
PROGRAM helloworld
   USE mpi_f08
   IMPLICIT none

   DOUBLE PRECISION :: my_data, my_result ! application-related data

! Turn it into an MPI program

! Experiment with Hello message in all processes and only in the master process

   WRITE (*,*) 'Hello World'

! Print out various information about MPI infrastructure

! Initialize different data in every MPI process depending on their rank

   WRITE(*,'(A,F9.2)') &
  & 'I am process ... with data = ', my_data

! Don't do - Broadcast data from some process to all other processes

! Don't do - Do some work with data (get result) and print data and result in every process

END PROGRAM 
